### ISSUE TITLE
listener conflicts related to headless services (warnings, errors, etc.)

### ISSUE THREAD

**Issue by Stono** [istio/istio#5992]
Hi, 
I'm seeing the following error in istio-pilot/discovery:

```
2018-06-03T00:27:50.659513Z     warn    buildSidecarOutboundListeners: listener conflict (TCP current and new TCP) on 0.0.0.0:9094, destination:outbound|9094||prometheus-alertmanager.istio-system.svc.cluster.local, current Listener: (0.0.0.0_9094 name:"0.0.0.0_9094" address:<socket_address:<address:"0.0.0.0" port_value:9094 > > filter_chains:<filters:<name:"envoy.tcp_proxy" config:<fields:<key:"deprecated_v1" value:<bool_value:true > > fields:<key:"value" value:<struct_value:<fields:<key:"route_config" value:<struct_value:<fields:<key:"routes" value:<list_value:<values:<struct_value:<fields:<key:"cluster" value:<string_value:"outbound|9094||alertmanager.data-platform.svc.cluster.local" > > > > > > > > > > fields:<key:"stat_prefix" value:<string_value:"outbound|tcp|9094" > > > > > > > > deprecated_v1:<bind_to_port:<> > )
```

For context, we have two teams running alertmanager, **in separate namespaces**.

  1. alertmanager.data-platform.svc.cluster.local
  2. prometheus-alertmanager.istio-system.svc.cluster.local

It appears when I use the same port section in the service for each of these, eg:

```
spec:
  ports:
  - name: cluster
    port: 9094
    protocol: TCP
    targetPort: cluster
```

If I rename the port in one of the second services to something else, everything starts working again and the errors in istio pilot go away.

EG: 
```
spec:
  ports:
  - name: cluster
    port: 9094
    protocol: TCP
    targetPort: cluster
```

Adding a prefix resolves it:
```
spec:
  ports:
  - name: http-cluster
    port: 9094
    protocol: TCP
    targetPort: cluster
```

**Comment by Stono** [comment]
Apologies, the late night must have gotten to me, the information above isn't quite right.

Having a port without a prefix (so treated as TCP) results in the error above, eg:

```
spec:
  ports:
  - name: frontend
    port: 9093
    protocol: TCP
    targetPort: frontend
```

Adding a prefix resolves it:
```
spec:
  ports:
  - name: http-frontend
    port: 9093
    protocol: TCP
    targetPort: http-frontend
```

So it seems to be specific to tcp services

**Comment by sakshigoel12** [comment]
@nmittler @rshriram can you take a look at this?

**Comment by rshriram** [comment]
Need config info

**Comment by mandarjog** [comment]
@Stono `https://github.com/istio/istio/blob/master/bin/dump_kubernetes.sh -z`   will gather config / logs etc.

**Comment by Stono** [comment]
Hey, 
I've written these manifests which you can use to recreate the problem..

```
apiVersion: v1
kind: Namespace
metadata:
  name: namespace1
---
apiVersion: v1
kind: Service
metadata:
  name: service1
  namespace: namespace1
spec:
  clusterIP: None
  ports:
  - name: someport
    port: 9999
    protocol: TCP
    targetPort: somepodport
  selector:
    app: someapp
  sessionAffinity: None
  type: ClusterIP
---
apiVersion: v1
kind: Namespace
metadata:
  name: namespace2
---
apiVersion: v1
kind: Service
metadata:
  name: service2
  namespace: namespace2
spec:
  clusterIP: None
  ports:
  - name: someport
    port: 9999
    protocol: TCP
    targetPort: somepodport
  selector:
    app: someapp
  sessionAffinity: None
  type: ClusterIP
```

Gives you:
```
2018-06-05T18:54:34.904181Z     warn    buildSidecarOutboundListeners: listener conflict (TCP current and new TCP) on 0.0.0.0:9999, destination:outbound|9999||service2.namespace2.svc.cluster.local, current Listener: (0.0.0.0_9999 name:"0.0.0.0_9999" address:<socket_address:<address:"0.0.0.0" port_value:9999 > > filter_chains:<filters:<name:"envoy.tcp_proxy" config:<fields:<key:"deprecated_v1" value:<bool_value:true > > fields:<key:"value" value:<struct_value:<fields:<key:"route_config" value:<struct_value:<fields:<key:"routes" value:<list_value:<values:<struct_value:<fields:<key:"cluster" value:<string_value:"outbound|9999||service1.namespace1.svc.cluster.local" > > > > > > > > > > fields:<key:"stat_prefix" value:<string_value:"outbound|tcp|9999" > > > > > > > > deprecated_v1:<bind_to_port:<> > )
```

**Comment by Stono** [comment]
I cant no, as these are headless services for statefulsets

**Comment by jasmine-jaksic** [comment]
Going forward, please use istio-users@googlegroups.com or https://istio.rocket.chat if you have questions.

**Comment by Stono** [comment]
@jasmine-jaksic I feel this is a bug, not a question.

**Comment by rshriram** [comment]
if you have two headless services on same port, the collision is expected. However, network access to those headless services should continue to work. Is the network connectivity blocked?

**Comment by Stono** [comment]
But they're in different namespaces?  Surely we shouldn't expect constant error messages in the log about collisions for something which is a perfectly valid configuration?

**Comment by rshriram** [comment]
Are you confused about the warning or are is there an actual network connectivity issue? the warning in your case is benign. We setup envoy passthough listeners for headless services. So, the traffic for <anyIP>:someport will simply pass through envoy to the intended destination (irrespective of namespaces or not). The only glitch is that you will not see stats for the second service, due to the way the implementation is done.

Long story short: if headless services, you need to ensure that all headless services (irrespective of namespaces) have unique ports, if you want to get stats for each service.

**Comment by Stono** [comment]
We did observe connectivity issues but I'm unable to reproduce it now, so
the only issue at the moment is the constant log message and as you say, a
loss off stats for all but the first service.

FYI we run a large multi tenant cluster where people work in isolation in
their own namespaces.  Multiple people deploy things like alertmanager as a
statefulsets with a headless service.

Those teams may not communicate with each other and they shouldn't need
to.  Users shouldn't be needing to coordinate the use of unique port
numbers on headless services to unrelated teams to work around this just
because that's "how it is implemented".

Surely there is a more unique way to key this than IP:Port?

Karl

**Comment by nmittler** [comment]
@rshriram 

> Surely there is a more unique way to key this than IP:Port?

Any thoughts?

**Comment by rshriram** [comment]
Sorry I missed this. Is it possible to deploy istio per namespace ? The problem with headless services is one of scale. Accessing pods directly instead of service VIPs means we need to construct Envoy listeners for all possible pod IPs, which is not a scalable option. So, we resort to using a special passthrough listener in Envoy listening on the headless service port (0.0.0.0:someport) that forwards traffic for these ports to the IP that the client-app resolved by itself (pod IP). This is as low as things go (linux listen() call with INADDR_ANY on some port).

**Comment by Stono** [comment]
@rshriram no it is not possible unfortunately.  We have a centralised PaaS system where istio lives in a single namespaces, and teams deploy apps into further isolated namespaces.

We have 1 istio namespace and all going to plan, 200 application namespaces.

**Comment by nmittler** [comment]
@rshriram given that renaming one of the ports resolves the problem, wouldn't involving the namespace into the key be a reasonable fix?

**Comment by rshriram** [comment]
He is not just renaming. He is declaring them as headless HTTP services. For http, we allow any number of headless services on same port because we differentiate services by the virtual host name. So if headless services in all namespaces are http, then please name them as http. The problem should be resolved then.

**Comment by Stono** [comment]
They're not http though.  We have redis and rabbitmq which are both TCP
headless services used for cluster auto Discovery.

Karl

**Comment by sakshigoel12** [comment]
@costinm

**Comment by rshriram** [comment]
Yeah. You are out of luck for tcp. Communication should work. Warnings - ignore. Metrics - unfortunately you can’t get metrics for each service when multiple different headless tcp services share same port.

**Comment by Stono** [comment]
Right, so just to be clear what you're saying is that you accept there is a problem but you've no intention of addressing it?

**Comment by rshriram** [comment]
I would love to address it if there is a way. There is currently no way to address this problem in a scalable fashion, short of adding few hundred envoy listeners (one per pod+port in the headless service). At that point, performance will severely degrade and other things will become unstable.

**Comment by rshriram** [comment]
These are two different problems. This thread is about multiple headless services on same port. The problem and outage you reported is caused by *conflicting* services on same port (http and TCP). Please open a new issue with the bug report.

**Comment by Stono** [comment]
@rshriram has it crossed your mind, even for one second that consumers of Istio, that don't always understand the internals of the code base, report issues based on their perception of the world and then rely on "the experts" to help them identify the root cause.

This issue details my experiences, and the logs I see.   And as I said early on in the issue, I **did** see a service outage then too but couldn't quite identify what caused it.

In both circumstances, I see the same warnings in pilot, so I am obviously going to continue to talk on the same thread.

If you would like to break it up, please go ahead.  But stop being difficult and pedantic with someone who is just trying to help you find issues in a product before you fly your flagship "1.0 release".

**Comment by rshriram** [comment]
I certainly understand where you are coming from and the frustration you are going through. Sorry about the outage. To help other consumers like you with similar issues, scoping one issue to one problem is of great help. Especially when someone is searching for related issues. Feel free to report your findings here. I will open up new issues, so that we can refer to them in PRs. Thanks for the bug reports and tests.

**Comment by rshriram** [comment]
Looking closely at the problem, this conflict arises due to multiple protocols on same port with a wildcard listener. Even if we setup listeners on every service IP, the same problem will occur when there are two service entries with HTTP and TCP, pointing to two different services.
Put another way, the whole of Istio is built on the assumption of client-side load balancing, service discovery, etc. Headless services and stateful sets break that pattern by directly accessing a service instance, instead of accessing through a virtual IP.

There are a few ways to solve this IMO:

1. Enforce same protocol for all services on same port
2. Enforce unique ports for each service

Either of these options requires a CI/CD or some form of admission control system that can detect port conflicts and prevent the services from being deployed in Kubernetes (and give a helpful error message to the end user). @ayj is it possible to write such an admission control plugin for services? Or is it restricted to just CRDs ?

**Comment by ayj** [comment]
Admission control isn't limited to CRD. Detecting conflicts within the same resource (i.e. service) is possible. However, we can't reliability enforce constraints across resources instances at creation time. We'd need to detect conflicts at runtime (best effort) and reflect that back to the user.

**Comment by Stono** [comment]
Hey, for what it's worth I was chatting to @rshriram on slack and at the least a "first come first serve" kind of approach would be preferable.

IE the latest deployed service is the one that is impacted by the conflict, not the first:

a. service a deploys http
b. service b deploys conflicting port
c. service b doesn't work, pilot shows errors/warnings 

My issue comes with service a breaking as a result of b being deployed at a later date

**Comment by nmittler** [comment]
@Stono @rshriram @costinm 

#7043 has been merged. I've been having trouble coming up with an integration test that recreates the original log message (I've been using the [simple example provided by @Stono above](https://github.com/istio/istio/issues/5992#issuecomment-394821567).

I found that #6402 changed the behavior of the listener conflict logic for TCP, meaning that post-0.8 the conflict will no longer be logged/reported. I suggested reverting the changed line...

In Slack, @rshriram said: 

> you can’t revert that line. It will break other TCP use cases.
but as I mentioned in that issue, two headless services on same port - connectivity will still work. Metrics won’t. because headless services are established on 0.0.0.0:port with original dst cluster. So, any connection to that port will go to the requested destination..

WRT @Stono's outage, connectivity was clearly an issue.  With the post-0.8 logic, two headless TCP services will no longer be found in conflict with each other, with the assumption @rshriram stated above.

I'm concerned that the current code will not avoid future outages ... or am I missing something?

**Comment by nmittler** [comment]
I think I have a better understanding now.  It was working, however it works in a different way than it did in 0.8.  In 1.0, the filtering occurs at the filterchain level, not at the listener level.

I've raised #7113 to improve the logging and to add a working e2e test.

**Comment by nmittler** [comment]
Quick update, the following have been submitted related to this issue:

#7101 (TCP routing cleanups/bug fixes: part 1)
#7113 (Improve outbound listener conflict messages)
#7160 (Adding unit test for TCP over TCP conflicts.)

The last piece an e2e test which is out for review (#7173).  Once that's in, we can mark this as fixed.

**Comment by nmittler** [comment]
#7173 is in. Marking this as fixed for now.
